iavl/iavl_proof_test.go

```diff
@@ -55,33 +65,424 @@ func TestIAVLTreeKeyExistsProof(t *testing.T) {
 	}
 
 	// query random key fails
-	_, proof, _ = tree.getWithKeyExistsProof([]byte("foo"))
+	_, proof, _ = tree.getWithProof([]byte("foo"))
 	assert.Nil(t, proof)
 
 	// query min key fails
-	_, proof, _ = tree.getWithKeyExistsProof([]byte{0})
+	_, proof, _ = tree.getWithProof([]byte{0})
 	assert.Nil(t, proof)
 
 	// valid proof for real keys
 	root := tree.Hash()
 	for _, key := range keys {
-		value, proof, _ := tree.getWithKeyExistsProof(key)
+		value, proof, _ := tree.getWithProof(key)
 		assert.NotEmpty(t, value)
 		if assert.NotNil(t, proof) {
 			err := proof.Verify(key, value, root)
-			assert.Nil(t, err, "%+v", err)
+			assert.NoError(t, err, "%+v", err)
 		}
 	}
+	// TODO: Test with single value in tree.
 }
 
-func TestIAVLTreeKeyNotExistsProof(t *testing.T) {
+func TestIAVLTreeKeyRangeProof(t *testing.T) {
 	var tree *IAVLTree = NewIAVLTree(0, nil)
+	require := require.New(t)
+	keys := [][]byte{}
+	for _, ikey := range []byte{
+		0x0a, 0x11, 0x2e, 0x32, 0x50, 0x72, 0x99, 0xa1, 0xe4, 0xf7,
+	} {
+		key := []byte{ikey}
+		keys = append(keys, key)
+		tree.Set(key, []byte(randstr(8)))
+	}
+	root := tree.Hash()
+
+	cases := []struct {
```

Comment:
AWESOME!

Can we get a second example of  `Sub-range, partially-existing keys, both directions` where left key exists and right doesn't,  and a second one of  `Overlapping range, both directions` where its out of range on the left ? 

Also maybe its worth checking that if we tamper with the proof in various ways the verification fails!